drivers/platform/stm32/stm32_irq.c

```diff
@@ -0,0 +1,267 @@
+#include "gpio.h"
+#include "stm32_gpio.h"
+
+/**
+ * @brief STM32 platform specific IRQ platform ops structure
+ */
+const struct irq_platform_ops stm32_irq_platform_ops = {
+	.init = &stm32_irq_ctrl_init,
+	.register_callback = &stm32_irq_register_callback,
+	.unregister = &stm32_irq_unregister,
+	.global_enable = &stm32_irq_global_enable,
+	.global_disable = &stm32_irq_global_disable,
+	.enable = &stm32_irq_enable,
+	.disable = &stm32_irq_disable,
+	.remove = &stm32_irq_ctrl_remove
+};
+
+/**
+* @struct stm32_irq_callback_desc
+* @brief Structure holding the callback functions for STM32 irqs
+* @note  The callback functions are mapped with 'irq_id' structure
+*/
+typedef struct stm32_irq_callback_desc {
+	uint16_t GPIO_Pin;
+	struct callback_desc desc;
+} stm32_irq_callback_desc[NB_INTERRUPTS];
+
+stm32_irq_callback_desc stm32_irq_callback;
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief STM32 GPIO external interrupt callback function
+ * @param GPIO_Pin - GPIO pin
+ * @return none
+ */
+void HAL_GPIO_EXTI_Callback(uint16_t GPIO_Pin)
+{
+	enum irq_id id;
+
+	/* Invoke the callback function attached to corresponding GPIO pin */
+	for (id = EXTERNAL_INT_ID1; id <= EXTERNAL_INT_ID5; id++) {
+		if (stm32_irq_callback[id].GPIO_Pin == GPIO_Pin) {
+			stm32_irq_callback[id].desc.callback(stm32_irq_callback[id].desc.ctx,
+							     id, NULL);
+			return;
+		}
+	}
+}
+
+/**
+ * @brief Initialize the STM32 interrupt
+ * @param desc - IRQ descriptor
+ * @param param - IRQ initialization parameters
+ * @return SUCCESS in case of success, FAILURE otherwise
+ */
+int32_t stm32_irq_ctrl_init(struct irq_ctrl_desc **desc,
+			    const struct irq_init_param *param)
+{
+	struct irq_ctrl_desc *descriptor;
+	struct stm32_irq_desc *extra;
+	int32_t ret;
+
+	if (!desc || !param)
+		return -EINVAL;
+
+	descriptor = (struct irq_ctrl_desc *)malloc(sizeof(*descriptor));
+	if (!descriptor) {
+		ret = -ENOMEM;
+		goto error;
+	}
+
+	extra = (struct stm32_irq_desc*)malloc(sizeof(*extra));
+	if (!extra) {
+		ret = -ENOMEM;
+		goto error;
+	}
+
+	switch (param->irq_ctrl_id) {
+	case EXTERNAL_INT_ID1:
+	case EXTERNAL_INT_ID2:
+	case EXTERNAL_INT_ID3:
+	case EXTERNAL_INT_ID4:
+	case EXTERNAL_INT_ID5:
+		extra->gpio_num = ((stm32_irq_init_param *)(param->extra))->gpio_num;
+		break;
+
+	default:
+		return -EINVAL;
+	}
+
+	extra->irq_type = ((stm32_irq_init_param *)(param->extra))->irq_type;
+
+	descriptor->extra = extra;
+	*desc = descriptor;
+
+	return SUCCESS;
+error:
+	free(extra);
+	free(desc);
+
+	return ret;
+}
+
+/**
+ * @brief Free the resources allocated by irq_ctrl_init()
+ * @param desc - IRQ descriptor
+ * @return SUCCESS in case of success, FAILURE otherwise
+ */
+int32_t stm32_irq_ctrl_remove(struct irq_ctrl_desc *desc)
+{
+	if (desc) {
```

Comment:
i would simplify this by doing:
```
if (!desc)
    return -EINVAL
```

Then the normal behavior code with return SUCCESS at the end.